### PR TITLE
Multi-Scale Hierarchical Slice Attention: coarse-to-fine cross-scale decoding

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The prior Multi-Scale Slice Attention attempt (PR #2373) failed because it added coarse and fine slice outputs **additively**, creating gradient interference between scales. The correct formulation is **hierarchical**: first compute coarse-scale slice tokens (16 large physics zones) to build global inter-foil context, then refine with fine-scale slices (96 local physics patches) that attend to the coarse tokens as additional context.

This mirrors the Swin Transformer's shifted-window hierarchy but in physics-state space. The coarse slices capture **tandem inter-foil coupling** (the main source of p_tan error), while fine slices capture local pressure peaks at leading/trailing edges. By making fine-scale attention conditioned on coarse-scale context (concatenated K,V), we avoid the additive gradient interference that killed PR #2373.

Key difference from PR #2373: **sequential coarse→fine with cross-scale attention**, not parallel additive.

Reference: Swin Transformer (Liu et al., ICCV 2021) — shifted window hierarchy. Also: Multi-Scale Vision Transformers (Fan et al., ICCV 2021) — pooling attention across scales.

## Instructions

**Changes to `cfd_tandemfoil/train.py`:**

Add flags: `--multiscale_hierarchical` and `--coarse_slices 16`

### 1. Modify `Physics_Attention_Irregular_Mesh` class

When `--multiscale_hierarchical` is enabled, modify **only the LAST TransolverBlock** (layer index `n_layers - 1`). Earlier layers remain unchanged with standard 96-slice attention.

In the last TransolverBlock's `Physics_Attention_Irregular_Mesh`:

**Stage 1 — Coarse attention (16 slices):**
```python
# Add a separate coarse slice projection
self.in_project_slice_coarse = nn.Linear(self.dim_head, self.coarse_slices)  # [D] -> [16]

# Compute coarse slice tokens
slice_weights_coarse = F.softmax(self.in_project_slice_coarse(x_reshape), dim=1)  # [B, H, N, 16]
slice_tokens_coarse = torch.einsum('bhns,bhnd->bhsd', slice_weights_coarse, x_reshape)  # [B, H, 16, D]

# Standard self-attention on coarse tokens
coarse_q = coarse_tokens  # or project via separate Q
coarse_k = coarse_tokens
coarse_v = coarse_tokens
coarse_attn = F.softmax(coarse_q @ coarse_k.transpose(-2, -1) / sqrt(D), dim=-1)
coarse_out = coarse_attn @ coarse_v  # [B, H, 16, D]
```

**Stage 2 — Fine attention with cross-scale context (96 slices):**
```python
# Standard fine slice tokens (existing 96 slices)
slice_weights_fine = F.softmax(self.in_project_slice(x_reshape), dim=1)  # [B, H, N, 96]
slice_tokens_fine = torch.einsum('bhns,bhnd->bhsd', slice_weights_fine, x_reshape)  # [B, H, 96, D]

# Cross-scale attention: fine Q attends to concatenated [coarse_out; fine] K,V
fine_q = slice_tokens_fine  # [B, H, 96, D]
cross_kv = torch.cat([coarse_out, slice_tokens_fine], dim=2)  # [B, H, 112, D]
fine_k = cross_kv
fine_v = cross_kv
cross_attn = F.softmax(fine_q @ fine_k.transpose(-2, -1) / sqrt(D), dim=-1)
fine_out = cross_attn @ fine_v  # [B, H, 96, D]
```

Then project fine_out back to node space via the existing `out_project_slice` (unchanged).

### 2. Flag wiring

```python
parser.add_argument('--multiscale_hierarchical', action='store_true')
parser.add_argument('--coarse_slices', type=int, default=16)
```

Pass `multiscale_hierarchical` and `coarse_slices` to the TransolverBlock constructor. In the TransolverBlock, only the LAST layer (check layer index) should use the hierarchical attention. All other layers use standard attention.

### 3. Important details

- The coarse projection `in_project_slice_coarse` is a NEW nn.Linear — do NOT reuse the existing `in_project_slice`.
- Initialize the coarse projection with the same scheme as the fine projection (default PyTorch init is fine).
- The cross-scale K,V concatenation means fine queries see 112 keys (16 coarse + 96 fine). This is the key mechanism — it allows each fine slice to "see" the global coarse context.
- Keep the existing `out_project_slice` unchanged — it still projects from 96 fine slices back to node space.
- Do NOT apply hierarchical attention to early layers — only the last layer. This limits overhead and lets early layers build features normally.

### 4. Run command

```bash
cd cfd_tandemfoil && python train.py \
    --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
    --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first \
    --pressure_deep --residual_prediction --surface_refine --te_coord_frame \
    --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
    --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
    --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 \
    --vortex_panel_n 64 \
    --multiscale_hierarchical --coarse_slices 16 \
    --wandb_group multiscale-hier-attn
```

Run 2 seeds (42, 73) as standard.

## Baseline

Current baseline (PR #2357, 2-seed avg):
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| p_in | 11.872 | < 11.872 |
| p_oodc | 7.459 | < 7.459 |
| p_tan | 26.319 | < 26.319 |
| p_re | 6.229 | < 6.229 |

Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 --wake_angle_feature --vortex_panel_velocity --vortex_panel_scale 0.1 --vortex_panel_n 64`

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2388 because CLOSING: ANP merged (PR #2379) and the new baseline dropped p_tan from 26.3 to 10.8. Multi-Scale Hierarchical Attention was designed to improve cross-scale information flow, but the ANP cross-attention already provides effective cross-foil coupling. This experiment hasn't started training yet — reassigning askeladd to a new P0 idea (Local Re_x Feature) on the ANP baseline..
